pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,711 @@
+	return nil
+}
+
+func doDelete(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for uuid, row := range txn.etcdCache[ovsOp.Table] {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		ovsResult.Count = ovsResult.Count + 1
+		etcdDelRow(txn, ovsOp.Table, uuid)
+	}
+	return nil
+}
+
+/* wait */
+func preWait(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	if ovsOp.Timeout != 0 {
+		err := fmt.Errorf("only support timeout 0")
+		ovsResult.Error = err.Error()
+		return err
+	}
+	return preSelect(txn, ovsOp, ovsResult)
+}
+
+func doWait(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for _, row := range txn.etcdCache[ovsOp.Table] {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		ok, err = isRowSelectedByUntil(row, ovsOp.Columns, ovsOp.Rows, ovsOp.Until)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		ovsResult.Count = ovsResult.Count + 1
+	}
+	if ovsResult.Count == 0 {
+		err := errors.New("failed condition")
+		ovsResult.Error = err.Error()
+	}
+	return nil
+}
+
+/* commit */
+func preCommit(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	durable := false // FIXME: missing ovsOp.Durable
+	if durable {
+		err := errors.New("not supported")
+		ovsResult.Error = err.Error()
+		return err
+	}
+	return nil
+}
+
+func doCommit(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return nil
+}
+
+/* abort */
+func preAbort(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	err := errors.New("aborted")
+	ovsResult.Error = err.Error()
+	return err
+}
+
+func doAbort(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return nil
+}
+
+/* comment */
+func preComment(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return nil
+}
+
+func doComment(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	now := time.Now()
+	key := makeKey(txn.dbname, "_journal", now.String())
+	comment := "my transact comment" // FIXME: missing ovsOp.Comment
+	etcdOp := clientv3.OpPut(key, comment)
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+	return nil
+}
+
+/* assert */
+func preAssert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	id := "1" // FIXME: missing ovsOp.Id
+	key := makeKey(txn.dbname, "_lock", id)
+	etcdOp := clientv3.OpGet(key)
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+	return nil
+}
+
+func doAssert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	id := "1" // FIXME: missing ovsOp.Id
+	_, ok := txn.etcdCache["_lock"][id]
```

Comment:
it's wrong, the lock assert should be part of the etcd transaction condition

Reply:
will address this later